src/main/java/org/spongepowered/api/item/inventory/InventoryArchetypes.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.inventory;
+
+import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+
+/**
+ * An enumeration of {@link InventoryArchetype}s.
+ */
+public class InventoryArchetypes {
+
+    // Base Archetypes
+    // TODO how to start?
+    public static final InventoryArchetype SLOT = DummyObjectProvider.createFor(InventoryArchetype.class, "SLOT");
+    public static final InventoryArchetype MENU_ROW = DummyObjectProvider.createFor(InventoryArchetype.class, "MENU_ROW");
```

Comment:
That makes more sense, +1 for providing a sensible rationale. To be honest maintaining parity with the text-based menus is sufficient reason to use callback instead, I'd rather have consistency any day of the week.

The listener can't really be passed to the archetype because the archetype may be consumed by another plugin and *that* plugin needs a way to supply appropriate listeners which is why I was thinking events made more sense (since event could just be registered on the archetype with an ID then consumed by *any* plugin) but thinking about it the event could simply be declared on the archetype and then attached to a listener at instance time:

```java
InventoryArchetype menuArchetype = .... ..... .withEvent("button_event_foo") .... .build();
Inventory menuInstance = Inventory.Builder.of(menuArchetype).listener("button_event_foo", Foo::bar).build();
```

> The event handlers in plugins are overhead, because you could implement a single one in the server impl that does the necessary button -> action mapping.

Not buying that at all.